### ISSUE TITLE
Add type generic to fetch()

### ISSUE THREAD

**Issue by ImRodry** [nodejs/undici#1198]
## This would solve...

The need to type `res.json() as ResponseType` on every json request which is usually not good practice in TS

## The implementation should look like...

When executing the fetch request, it would be nice to be able to pass an interface as the response type, for example:
```ts
fetch<MyResponseType>("https://example.com")
```

## I have also considered...

This type generic could also be placed on the `res.json()` function, as that should be the only one that needs this type, either way works just as fine. Both of these alternatives are better than the current one, which is doing `res.json() as MyResponseType`

## Additional context

This fix would be as simple as adding this to the typescript declaration file, so I'd be glad to open a PR about this if needed
![image](https://user-images.githubusercontent.com/38259440/152207270-7f39201f-6f9a-41d2-b7b5-c53af4ad24a9.png)

**Comment by KhafraDev** [comment]
This has been discussed a few times previously:
- https://github.com/nodejs/undici/pull/1048
- https://github.com/nodejs/undici/pull/997#discussion_r695334060

**Comment by ImRodry** [comment]
I think the argument that a type generic feels safer when it actually isn’t is just a misconception as a person who is familiar with TypeScript would know that it’s impossible for it to fetch an API and check if the type you provided is correct. In my opinion, if you’re casting the result of a fetch operation to a specific type it’s because you’re expecting that type to be like that and, if it isn’t, something probably went very wrong and you want to get the error so you can look further into it. After all, this would simply be more aesthetically pleasing and would have no negative consequences, as I’m sure no good TypeScript developer likes using the `as` keyword a lot.

**Comment by suneettipirneni** [comment]
TLDR; I agree with node-fetch's and Typescript's opinions, type casting should be explicit not implicit.

> I think the argument that a type generic feels safer when it actually isn’t is just a misconception as a person who is familiar with TypeScript would know that it’s impossible for it to fetch an API and check if the type you provided is correct. 

Generics are usually used where the type of the data can always be guaranteed *by the library*. In the world of HTTP there's really never a guarantee of the data you'll receive, so this type casting/narrowing should be up to the user, not the library.

If you look at the flow of types, the generic would just be acting like a cast for a given type. This isn't a great usage for a generic parameter as it should only be an indicator of bounds and/or types that 100% known during compile-time.

> In my opinion, if you’re casting the result of a fetch operation to a specific type it’s because you’re expecting that type to be like that and, if it isn’t, something probably went very wrong and you want to get the error so you can look further into it. After all, this would simply be more aesthetically pleasing and would have no negative consequences, as I’m sure no good TypeScript developer likes using the `as` keyword a lot.

This would have negative consequences because the actual type casting happens with the library, which gives the false sense that your result will always be of a given type, when in reality that can never be guaranteed. In addition, if you have complex responses you can use other narrowing mechanisms like `typeof`, `in`, or any user defined type-guards if using `as` doesn't feel safe enough.

**Comment by ImRodry** [comment]
> This would have negative consequences because the actual type casting happens with the library, which gives the false sense that your result will always be of a given type, when in reality that can never be guaranteed. In addition, if you have complex responses you can use other narrowing mechanisms like typeof, in, or any user defined type-guards if using as doesn't feel safe enough.

If you're using a consistent API the return type should always be the same and you should know what it is. If this isn't the case, you likely have an error in your code that you want to fix.
You can also set the type generic to a union of types and then perform whatever checks you need to assure which type you got on your API call. Having that done in a generic or through `as` is the same exact thing, except one is more aesthetically pleasing and readable than the other.

**Comment by suneettipirneni** [comment]
> You can also set the type generic to a union of types and then perform whatever checks you need to assure which type you got on your API call. Having that done in a generic or through `as` is the same exact thing, except one is more aesthetically pleasing and readable than the other.

This still doesn't change the fact that generics shouldn't be used as casting mechanism and that the casting shouldn't be handled by the library. Typescript ships with proper  casting mechanisms for these very situations. I don't think aesthetics should override these more proper ways of type casting.

**Comment by ImRodry** [comment]
I really don't understand how allowing the user to specify a type generic is putting any responsibility on the library, since there's not even any type that the generic must extend, so it's really just opening the possibility for a better experience on the user's side, with no downsides to the library itself nor the users who don't wanna use this feature. If you prefer using `as`, you can keep doing so, otherwise you can use the generic

**Comment by suneettipirneni** [comment]
> I really don't understand how allowing the user to specify a type generic is putting any responsibility on the library, since there's not even any type that the generic must extend, so it's really just opening the possibility for a better experience on the user's side, with no downsides to the library itself nor the users who don't wanna use this feature. If you prefer using `as`, you can keep doing so, otherwise you can use the generic

The comment by the typescript team member in the provided comment answers this quite adequately:

> https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36868#issuecomment-512989650
